components/theia/packages/gitpod-extension/src/browser/gitpod-user-storage-provider.ts

```diff
@@ -93,6 +94,37 @@ export class GitpodUserStorageProvider implements FileSystemProviderWithFileRead
         }]);
     }
 
+    // To avoid having badly written extensions executing too much requests and putting pressure on other components we
+    // cache writes here.
+    protected readonly contentCache = new Map<string, string>();
+    protected readonly updatedSinceLastWrite: string[] = [];
+    protected async cachedWrite(resource: GitpodServer.UpdateUserStorageResourceOptions): Promise<void> {
+        this.contentCache.set(resource.uri, resource.content);
+        this.updatedSinceLastWrite.push(resource.uri);
+    }
+
+    protected async cachedRead(req: GitpodServer.GetUserStorageResourceOptions): Promise<string> {
+        const cached = this.contentCache.get(req.uri);
+        if (cached !== undefined) {
+            return cached;
+        }
+        return this.getServer().getUserStorageResource(req);
+    }
+
+    protected startPeriodicCacheWriter(interval: number) {
+        setInterval(async () => {
+            const server = this.getServer();
+            
+            let uri = this.updatedSinceLastWrite.pop();
```

Comment:
depending on how frequent write calls are done and how slow the actual write here is, we could end up in multiple (potentially endless) writes for the same URI. I think we should dereference the array to a local var and assign an empty array instead of the pop() business.

Reply:
I see what you mean, but a local array just trades it against potential lost updates. Instead chained `setTimeout` should fix this.
I originally decided against it because I deemed the timings to be way off, but can easily change that.

Reply:
Why would you get lost updates? It's node so there is no concurrence. The reassignment is atomic.

Reply:
> It's node so there is no concurrency

Thanks for that reminder.... :wink: 

> Why would you get lost updates?

What I meant was: The old version naively used `setInterval`, which might lead to situations where two `thunks` race for the same `uri`. Which would not have lead to lost updates (my bad), but too many requests.

Anyways, the current code fixes all of the issues mentioned.